Migrate all tests

ronyeh opened the pull request:
I have transitioned `flow.html` to only include `vexflow-tests.js`, which is bundled by webpack (and includes any VexFlow library classes that are imported by the tests).

My plan is to drop a `/* eslint-disable */` at the top of each test file for now. Once every test is renamed to *.ts, then we can go back and actually add types to enforce type checking and discover bugs.

This PR subsumes https://github.com/0xfe/vexflow/pull/1065 so we can ignore https://github.com/0xfe/vexflow/pull/1065.

ronyeh commented:
A big milestone!

All the tests have been migrated to *.ts files. Everything is webpack bundled into a single `vexflow-tests.js`, and flow.html now only includes one copy of VexFlow.

All tests under `flow.html` pass.

For now, I put this at the top of each tests file to hide any eslint / ts issues:

```
/* eslint-disable */
// @ts-nocheck
```

After we eventually merge this PR, we will need to revisit each tests file and remove those two lines to see what issues crop up.

The most common issue I see is that tests will omit parameters, but our current TypeScript-ified API declares those parameters are required. We will need to revisit those cases to see which params can be made optional via param? or via  param=default_value.

There are still some open issues with this PR, because the command line scripts need to be updated.

rvilarl commented on src/easyscore.ts:
```diff
@@ -15,6 +15,7 @@ import { RenderContext } from './types/common';
 import { Accidental } from './accidental';
 import { Modifier } from './modifier';
 import { Voice } from './voice';
+import { Note } from 'note';
```
change to `from './note'`

ronyeh replied:
While migrating tests, I noticed that I could omit paths for any class under src/.

For example:
```
import { EasyScore } from '../src/easyscore';
import { StaveNote } from '../src/stavenote';
import { System } from '../src/system';
```
can be rewritten as
```
import { EasyScore } from 'easyscore';
import { StaveNote } from 'stavenote';
import { System } from 'system';
```

I think this is due to the baseURL field in the tsconfig.

If VS Code and TS compiler and webpack don't complain, and everything builds fine, does that mean we can omit all the ./ in front of all the imports?

We can do a separate PR for this. I think it would make everything look slightly cleaner if we can omit all the starting ./

0xfe replied:
Yep, agreed.

rvilarl commented on src/staveline.ts:
```diff
@@ -154,7 +154,9 @@ export class StaveLine extends Element {
-  protected notes: StaveLineNotes;
+
+  // Initialized by this.setNotes(notes) in the constructor.
+  protected notes!: StaveLineNotes;
```
The linter does not complain?

ronyeh replied:
Surprisingly, I am not getting any eslint complaints when I add a ! 
I need to investigate...

rvilarl commented on src/tables.ts:
```diff
@@ -1308,7 +1305,7 @@ export const Tables = {
             rest: true,
             position: 'B/4',
-            dot_shiftY: 1.5,
+            dot_shiftY: -2.5,
```
This change does not affect the output of the tests?

ronyeh replied:
We never tested 128th rests before. When I added a 128th rest to the test cases, I noticed the dot was in the wrong place.

0xfe replied:
Good catch!

rvilarl commented on src/tabslide.ts:
```diff
@@ -26,18 +26,19 @@ export class TabSlide extends TabTie {
   }
 
+  /**
+   * @param notes is a struct that has:
```
you could comment each attribute
@param notes.first_note 
...
if you wanted

rvilarl commented on src/tabtie.ts:
```diff
@@ -16,18 +16,18 @@ export class TabTie extends StaveTie {
   }
 
+  /**
+   * @param notes is a struct that has:
```
as above, you could comment each attribute if you wanted.

rvilarl commented on src/tuplet.ts:
```diff
@@ -73,7 +73,7 @@ export class Tuplet extends Element {
   protected x_pos: number;
   protected width: number;
-  protected location: number;
+  protected location!: number;
```
The linter does not comply?

ronyeh replied:
It seems like adding a ! is equivalent to asking the TS compiler to avoid flagging this line. It's basically like saying "trust me, I know what I am doing here because I actually do initialize this property later".

eslint does not complain. Perhaps there is a setting we can turn on if we care to.

ronyeh replied:
More info here:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#definite-assignment-assertions

We definitely need to be careful when using the **definite assignment** operator. (It's similar to the non-null assertion, which the eslint will definitely complain about.)

For now, we can just require a comment above the instance variable that indicates where it is assigned. In this case, I will need to add a comment to tuplet.ts to show where `location` is assigned.

ronyeh replied:
It's actually easy to search for where we use this feature.

Doing a project wide search for the string `!:` I see that I have 40 instances of it. The master version has 34. So I will need to check where I added them and provide comments to verify that those instance variables are definitely assigned via a helper function.

rvilarl commented:
@0xfe I have review the code change, I suggest to accept the test changes without review on the basis of identical output, otherwise the revisison would take too much effort.
@ronyeh this is great.

ronyeh commented:
Ready for comments and code review!

ronyeh commented:
To do a sanity check on the old API (e.g., `new VF.Formatter()`), I added some new example test cases. As we remove more VF.* prefixes, we can add more of these checks along the way.

```
test('VF.* API', this.VF_Prefix);


// Use QUnit to verify that the classes are identical.
VF_Prefix(): void {
  equal(Voice, VF.Voice);
  equal(Registry, VF.Registry);
  propEqual(new Formatter(), new VF.Formatter(), 'new Formatter()');
  equal(StaveNote, VF.StaveNote);
  notEqual(Stave, VF.StaveNote);
  equal(Stave, VF.Stave);
  equal(Beam, VF.Beam);
  equal(Note, VF.Note);
  equal(Bend, VF.Bend);
  equal(Annotation, VF.Annotation);
  equal(StaveConnector, VF.StaveConnector);
},
```

ronyeh commented:
There are 4 tests with visual diffs. 3 of the diffs are due to my small changes and bug fixes. 1 of the diffs was unexpected, but I believe the output is now correct, whereas the output was previously incorrect. (So maybe I fixed a small bug somewhere?)

I will put together some before and after GIFs.

Visual Diffs: 
```
Rests.Rests___Dotted.*.png
This was expected because I added a 128th rest to the test case, and also fixed the dot's vertical location.

KeySignature.Key_Signature_Change_test.*.png
This was expected. The width of the stave changed, because I fixed a bug in the test.
The previous test code was buggy. It called `factory.Stave(10, 10, 800)` even though `factory.Stave(options)` only accepts 1 param.

StaveModifier.Vertical_Bar_Test.*.png
The previous test output was incorrect. The current test output is correct. However, this diff was not expected. I may have fixed a tiny bug somewhere as I ported the tests. I might have to investigate this one.

StringNumber.Multi_Voice_With_Strokes__String___Finger_Numbers.*.png
I changed the order in which the notes/voices were drawn, so that the ledger lines from the second voice do not display on top of the note heads from voice 1. While fixing this test case, I discovered that VexFlow draws TWO ledger lines for middle C when two separate voices both require a middle C ledger line.



```

ronyeh commented:
I understand this is a pretty big PR :-) and I would like to make sure we are extremely confident that it doesn't introduce bugs before we merge. (In fact I believe it fixes a few old bugs, and exposes TS migration issues that we will need to address.)

I think a good first step will be for Mohit and Rodrigo to directly git clone my branch:

https://github.com/ronyeh/vexflow/tree/migration/tests

Then npm install and run grunt and poke around through all the test files.

You will see that the work is not complete. There are still many instances of VF.* that we need to remove. There are also eslint-disable and @ts-nocheck at the top of each file. If you temporarily remove those, you will see the TypeScript issues that pop up. The most common one is where the test case omits a param, but the TS method signature declares that the param is required. In many cases, the simplest fix is to add a ? to the method param (after verifying that the method does provide a default value if the param is undefined).

My next step will be to investigate the one unexpected visual diff (I believe the master revision had buggy output).

I will also try to put together a list of any small bugs I squashed, and add comments to any place I added the ! definite assignment assertion operator.

ronyeh commented:
Here are the six ! operators that I added. I made sure that all of them have comments. There are already other uses of this operator in the code base. We can add more comments next time.

Several of the ! operators were added to the `notes` instance variable. I discovered an issue where `TabSlide` extends `TabTie` extends `StaveTie`, but all three constructors called `this.setNotes()`. So creating a new `TabSlide` would call setNotes() three times (via super()).

```
element.ts
  // fontStack and musicFont are both initialized by the constructor via this.setFontStack(...).
  protected fontStack!: Font[];
  protected musicFont!: Font;

tuplet.ts
  // location is initialized by the constructor via setTupletLocation(...).
  protected location!: number;

stavetie.ts
  // notes is initialized by the constructor via this.setNotes(notes).
  protected notes!: TieNotes;

stavehairpin.ts
  // notes is initialized by the constructor via this.setNotes(notes).
  protected notes!: Record<string, Note>;

staveline.ts
  // These five instance variables are all initialized by the constructor via this.setNotes(notes).
  protected notes!: StaveLineNotes; // <= I added this !
  protected first_note!: StaveNote;
  protected first_indices!: number[];
  protected last_note!: StaveNote;
  protected last_indices!: number[];
```

rvilarl commented:
Fixes #1079

ronyeh commented:
![A](https://user-images.githubusercontent.com/239113/127761836-8c145844-5f9a-4463-a02f-b61d7c5cd232.gif)

This was one of the four visual diffs: `StaveModifier.Vertical_Bar_Test.*.png`

The reference test output was incorrect. The current test output (with two visible lines) is correct. 

This diff was NOT expected. I will try to investigate what small change I made that resulted in this bug fix.

rvilarl commented:
> I think a good first step will be for Mohit and Rodrigo to directly git clone my branch:
> 
> https://github.com/ronyeh/vexflow/tree/migration/tests
> 
> Then npm install and run grunt and poke around through all the test files.

@ronyeh I cloned your branch and used `npm run test:reference`to get the differences on the tests output compared to master. Here they are in pairs (first this branch, second master), only 4 differences:

1) KeySignature Key_Signature_Change_test Bravura
![KeySignature Key_Signature_Change_test Bravura_Current](https://user-images.githubusercontent.com/22865285/127761600-a8d85fd5-b4f9-46ff-a499-47080c2075bb.png)

![KeySignature Key_Signature_Change_test Bravura_Reference](https://user-images.githubusercontent.com/22865285/127761606-7d7db0fc-d5ef-4aa3-8ffd-dd1e2c6b484d.png)

2) Rests Rests___Dotted Bravura **(128r was added, this is fine)**
![Rests Rests___Dotted Bravura_Current](https://user-images.githubusercontent.com/22865285/127761627-8de24141-3cb3-4491-a0c8-d7c486872a2a.png)

![Rests Rests___Dotted Bravura_Reference](https://user-images.githubusercontent.com/22865285/127761645-25396a5a-5ca0-4f31-aa8e-d16b2ffdda16.png)

3) StaveModifier Vertical_Bar_Test Bravura 
![StaveModifier Vertical_Bar_Test Bravura_Current](https://user-images.githubusercontent.com/22865285/127761655-08057dd3-e091-45fb-929a-e68dcfced9b3.png)

![StaveModifier Vertical_Bar_Test Bravura_Reference](https://user-images.githubusercontent.com/22865285/127761668-40623efa-0073-4970-9d77-4211d22498dc.png)

4) StringNumber Multi_Voice_With_Strokes__String___Finger_Numbers 
![StringNumber Multi_Voice_With_Strokes__String___Finger_Numbers Bravura_Current](https://user-images.githubusercontent.com/22865285/127761678-1a1d022a-28a8-43af-8bfb-8297ccb17e78.png)

![StringNumber Multi_Voice_With_Strokes__String___Finger_Numbers Bravura_Reference](https://user-images.githubusercontent.com/22865285/127761681-0f8c4f64-6454-4498-abeb-fb2648286873.png)

rvilarl commented:
> 
> This diff was NOT expected. I will try to investigate what small change I made that resulted in this bug fix.

I also realised that, we are both looking at the same thing :)

rvilarl commented:
> > This diff was NOT expected. I will try to investigate what small change I made that resulted in this bug fix.
> 
> I also realised that, we are both looking at the same thing :)

@ronyeh looking at the code I think that the current representation is the right one, you also commented on the ouput

```js
/**
   * Draw three vertical lines. The last line is not visible, because it overlaps
   * the right border of the stave at x = 300.
   */
  drawVerticalBar(options: TestOptions, contextBuilder: ContextBuilder): void {
    const ctx = contextBuilder(options.elementId, 400, 120);
    const stave = new Stave(10, 10, 300);
    stave.setContext(ctx);
    stave.draw();
    stave.drawVerticalBar(100);
    stave.drawVerticalBar(150);
    stave.drawVerticalBar(300);

    ok(true, 'all pass');
  },
```
![StaveModifier Vertical_Bar_Test Bravura_Current](https://user-images.githubusercontent.com/22865285/127761655-08057dd3-e091-45fb-929a-e68dcfced9b3.png)

ronyeh commented:
@rvilarl I figured it out!

The master branch is running the wrong test code:

https://github.com/0xfe/vexflow/blob/2315e7aba40a9bd0328ccb0b6c155eeff4489417/tests/stavemodifier_tests.js#L8-L12

ronyeh commented:
As you can see above, the master branch was calling .draw and .drawVerticalBar in the Stave tests, not the StaveModifier tests.

So when I converted it to TypeScript, it revealed the error and I pointed it to the draw and drawVertical bar defined in StaveModifierTests. I think it's as simple as that.

rvilarl commented:
@0xfe checked the test results are all fine! From my point of view, ready to merge.

0xfe commented on .vscode/settings.json:
```diff
@@ -1,6 +1,14 @@
 // Place your settings in this file to overwrite default and user settings.
 {
   "eslint.enable": true,
+  "files.exclude": {
```
Thanks for this. vscode-eslint starts to get laggy after a while.

0xfe commented:
Ron, this is fantastic work! Great job! 

Rodrigo, thanks for reviewing this. I agree that this is ready to merge. Absolutely thrilled at how far we've come, thanks to you two.
